crates/aptos-jwk-consensus/src/epoch_manager.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    certified_update_producer::RealCertifiedUpdateProducer,
+    jwk_manager::JWKManager,
+    network::{IncomingRpcRequest, NetworkReceivers, NetworkSender},
+    network_interface::JWKConsensusNetworkClient,
+    types::JWKConsensusMsg,
+};
+use anyhow::Result;
+use aptos_bounded_executor::BoundedExecutor;
+use aptos_channels::{aptos_channel, message_queues::QueueStyle};
+use aptos_config::config::IdentityBlob;
+use aptos_consensus_types::common::Author;
+use aptos_event_notifications::{
+    EventNotification, EventNotificationListener, ReconfigNotification,
+    ReconfigNotificationListener,
+};
+use aptos_logger::{error, info};
+use aptos_network::{application::interface::NetworkClient, protocols::network::Event};
+use aptos_reliable_broadcast::ReliableBroadcast;
+use aptos_types::{
+    account_address::AccountAddress,
+    epoch_state::EpochState,
+    jwks::{ObservedJWKs, ObservedJWKsUpdated, SupportedOIDCProviders},
+    on_chain_config::{
+        FeatureFlag, Features, OnChainConfigPayload, OnChainConfigProvider, ValidatorSet,
+    },
+};
+use aptos_validator_transaction_pool as vtxn_pool;
+use futures::StreamExt;
+use futures_channel::oneshot;
+use std::{sync::Arc, time::Duration};
+use tokio_retry::strategy::ExponentialBackoff;
+
+pub struct EpochManager<P: OnChainConfigProvider> {
```

Comment:
@grao1991 we care about epoch for jwk consensus because it has to be between validator-set members only, and validator set may change every epoch.

But it seems if a validator left validator set, the current implementation still allows it to participate in the next epoch, which seems a bug.

Reply:
I guess you only care about if a validator is in the validator set when aggregate the signature?